src/ir_Bosch.cpp

```diff
@@ -25,37 +27,298 @@ const uint16_t kBoschZeroSpace = 571;
+  if( nbytes % kBosch144BytesPerSection != 0 )return;
   // Set IR carrier frequency
-  enableIROut(kBoschFreq);
+  enableIROut( kBoschFreq );
 
-  for (uint16_t r = 0; r <= repeat; r++) {
-    const uint16_t kSectionByteSize = nbytes / kBosch144NrOfSections;
-    for (uint16_t offset = 0; offset < nbytes; offset += kSectionByteSize)
+  for( uint16_t r = 0; r <= repeat; r++ )
+  {
+    for( uint16_t offset = 0; offset < nbytes; offset += kBosch144BytesPerSection )
       // Section Header + Data + Footer
-      sendGeneric(kBoschHdrMark, kBoschHdrSpace,
-                  kBoschBitMark, kBoschOneSpace,
-                  kBoschBitMark, kBoschZeroSpace,
-                  kBoschBitMark, kBoschFooterSpace,
-                  data + offset, kSectionByteSize,
-                  kBoschFreq, true, 0, kDutyDefault);
-    space(kDefaultMessageGap);  // Complete guess
+      sendGeneric( kBoschHdrMark, kBoschHdrSpace,
+                   kBoschBitMark, kBoschOneSpace,
+                   kBoschBitMark, kBoschZeroSpace,
+                   kBoschBitMark, kBoschFooterSpace,
+                   data + offset, kBosch144BytesPerSection,
+                   kBoschFreq, true, 0, kDutyDefault );
+    space( kDefaultMessageGap );  // Complete guess
   }
 }
 
 #endif  // SEND_BOSCH144
 
+/// Class constructor.
+/// @param[in] pin GPIO to be used when sending.
+/// @param[in] inverted Is the output signal to be inverted?
+/// @param[in] use_modulation Is frequency modulation to be used?
+IRBosch144AC::IRBosch144AC( const uint16_t pin, const bool inverted, const bool use_modulation )
+    : _irsend( pin, inverted, use_modulation )
+    {
+      stateReset();
+    }
+
+/// Reset the internal state to a fixed known good state.
+void IRBosch144AC::stateReset( void ) 
+{
+  setRaw( kBosch144DefaultState );
+}
+
+/// Set up hardware to be able to send a message.
+void IRBosch144AC::begin( void )
+{
+  _irsend.begin(); 
+}
+
+#if SEND_BOSCH144
+/// Send the current internal state as an IR message.
+/// @param[in] repeat Nr. of times the message will be repeated.
+void IRBosch144AC::send( const uint16_t repeat )
+{
+  if( !powerFlag ) _irsend.sendBosch144( kBosch144Off, sizeof( kBosch144Off ), repeat ); // "Off" is a 96bit message
+  else _irsend.sendBosch144( getRaw(), kBosch144StateLength, repeat );
+  // other 96bit messages are not yet supported
+}
+#endif  // SEND_BOSCH144
+
+/// Get a copy of the internal state as a valid code for this protocol.
+/// @return A valid code for this protocol based on the current internal state.
+uint8_t* IRBosch144AC::getRaw( void ) 
+{ 
+  return _.raw; 
+}
+
+/// Set the internal state from a valid code for this protocol.
+/// @param[in] new_code A valid code for this protocol.
+void IRBosch144AC::setRaw( const uint8_t new_code[] ) 
+{
+  std::memcpy( _.raw, new_code, kBosch144StateLength );
+}
+
+void IRBosch144AC::setPower( const bool on )
+{
+  powerFlag = on;
+}
+
+bool IRBosch144AC::getPower( void ) const 
+{
+  return true;
```

Comment:
Shouldn't you check if the `_.raw`/state code isn't the Off message and if `powerFlag` is `True`?

Reply:
I guess I didn't realize then that I wasn't done yet.

Reply:
Well, this protocol seems more complicated than most. So, kind of expected.

Reply:
getPower() now queries the state of powerFlag.